## Correciones examen 1
### Ana Sofía Hinojosa Bale

In [261]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from scipy import stats
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
import statsmodels.api as sm

### Teoría

1.        ¿Qué es un pipeline?

Un pipeline es un diagrama que muestra de forma ordenada los pasos que se siguen en un proceso. En el pipeline se representan los datos que entran, las transformaciones o tareas que se realizan, y los resultados que se obtienen al final. Sirve para entender fácilmente cómo fluye la información y cómo se conectan las distintas etapas.

2.        ¿Cuál es el propósito de realizar regresiones? Explica las ventajas y desventajas de los dos planteamientos vistos en clase.
   
El enfoque de estadística inferencial busca entender el porqué, explicando las relaciones entre las variables y comprobando hipótesis para obtener conclusiones válidas, como en el análisis de significancia que se hace en el examen práctico. La ventaja de este enfoque es que permite comprobar supuestos y tomar decisiones con fundamento teórico. Sin embargo, como no se centra en la predicción, puede ser menos preciso para anticipar nuevos datos.

El enfoque de machine learning se orienta a obtener la mejor salida posible para predecir datos, ya sea completando valores faltantes en los que se tienen o anticipando datos futuros. Su ventaja es que suele ser muy preciso y optimiza los resultados. No obstante, en muchos casos también resulta importante comprender por qué ocurren las cosas, algo que la estadística inferencial aborda mejor.

3.        ¿En qué consiste el proceso de escalamiento de factores?

El escalamiento de factores consiste en transformar los valores de las variables para que todas estén en un mismo rango o intervalo de valores. De esta manera, ninguna variable tiene más peso que otra solo por el tamaño de sus valores, y además permite compararlas de forma sencilla y mejora el desempeño de los modelos que usan los datos.

4.        Explica el propósito de penalizar factores en una regresión.
   
El propósito de penalizar factores en una regresión es evitar que los coeficientes (las betas) se hagan demasiado grandes, ya sea reduciéndolos o incluso haciéndolos cero. Esto sirve para que el modelo no dependa solo de unas pocas variables, sino que tome en cuenta la información de manera más equilibrada, evitando que sólo se tome en cuenta cierta información en lugar de todos los factores.

5.        ¿Cuál es la relación entre escalamiento y penalización?

La relación entre escalamiento y penalización es que el escalamiento hace que las betas sean comparables entre sí. De esta manera, si algunas betas crecen demasiado, la penalización puede actuar correctamente. Si no se hace escalamiento, los coeficientes podrían ser grandes no porque el modelo dependa demasiado de esas variables, sino simplemente porque los datos no están en la misma escala, lo que distorsiona la interpretación y la penalización.

6.        Explica el concepto de una prueba de hipótesis.

Una prueba de hipótesis consiste en plantear una hipótesis nula, que generalmente dice que no hay efecto, que no existe relación entre variables, o que algo es igual a un valor específico. La hipótesis nula representa la situación que asumimos verdadera mientras no tengamos evidencia suficiente en contra. Por otro lado, la hipótesis alterna es la afirmación opuesta, se acepta si los datos muestran que la hipótesis nula probablemente no es correcta.

7.        Explica la interpretación de un p-value de una prueba de hipótesis que compara contra una media µ.

El p-value en una prueba de hipótesis que compara contra una media indica la probabilidad de obtener un resultado igual al planteado por la hipótesis nula, por ejemplo, que la beta sea igual a cero, lo que significaría que no hay relación entre la variable y la variable que se quiere predecir, asumiendo que la hipótesis nula es cierta. Mientras más pequeño sea el p-value, menos compatibles son los datos con la hipótesis nula, lo que sugiere que podría existir una relación real entre las variables.

Por ejemplo, cuando queremos ver si una variable es significativa en un modelo, la hipótesis nula dice que no hay relación entre esa variable y lo que queremos predecir, aunque pueda parecer que sí, mientras que la hipótesis alterna dice que sí existe esa relación. Si usamos un nivel de confianza del 95 por ciento, que corresponde a un nivel de significancia de 0.05, un p-value menor a 0.05 indica que los datos son poco compatibles con la hipótesis nula, por lo que se rechaza y se concluye que la variable tiene un efecto significativo.

8.        Describe el propósito de realizar cross-validation.

El propósito de realizar cross-validation es evitar que el modelo se sobreajuste a los datos que se han dado. Esto se hace separando los datos en conjuntos de entrenamiento y prueba, de manera que el modelo se entrene con los datos de entrenamiento y luego se evalúe con los datos de prueba. Así podemos verificar si el modelo realmente generaliza y funciona bien con datos nuevos, o si está sobreajustado y no será capaz de predecir correctamente información que no haya visto antes.

9.        Describe los pasos que seguirías al hacer un análisis exploratorio de datos. Justifica cada paso.
    
Los pasos que seguiría al hacer un análisis exploratorio de datos sería:

- Ver el dataset. Esto para saber con qué variables cuento y ver, aunque no a profundidad, el tipo de valores que tendré.
- Sacar el resumen estadístico del dataset. Esto para darme una idea general de los datos que tengo, de qué a qué valores llegan.
- Ver los tipos de variables que tengo. Esto para ver si tengo variables categóricas o numéricas y saber si voy a tener que hacer variables dummy.
- Ver si hay valores faltantes en las variables. Esto para ver si voy a tener que quitar dichos valores faltantes del dataset.

10.        ¿Qué es el teorema del límite central?

El teorema del límite central establece que, si se toman muestras de tamaño suficientemente grande de cualquier población, las medias de esas muestras se distribuirán aproximadamente de forma normal, es decir, en forma de campana, sin importar cómo esté distribuida la población original.

# 1

Encontrar la parábola que pasa exactamente por los puntos (1, 2), (2,1) y (3,2). Utiliza la librería sklearn.

In [262]:
x = np.array([1,
              2,
              3])
y = np.array([2,
              1,
              2])

In [263]:
x_2 = np.vstack((x, x**2)).T

lr = LinearRegression()
lr.fit(x_2, y)

LinearRegression()

In [264]:
b0, b1, b2 = lr.intercept_, lr.coef_[0], lr.coef_[1]
print(f"Parábola: y = {b0:.2f} + {b1:.2f}x + {b2:.2f}x^2")

Parábola: y = 5.00 + -4.00x + 1.00x^2


# 2

Primero, se importan las librerías necesarias para que el resto del código funcione.

Después, se guardan los datos en una variable llamada housing en un data frame, el que después se divide en x (los datos con los que se va hará el modelo), y y (los datos que el modelo busca pronosticar). 

Después de separar los datos en x y y, se llama a la función StandarScaler(), numerical_transformer, y a las columnas del dataframe donde vienen las variables x, se les guarda en una variable llamada numerical_features. Con esto, se crea preprocessor, que es donde después se escalan con la función definida los datos de cada una de dichas columnas.

Después se guarda la funcion LinearRegression() en la variable model para poder llamarla facilmente y se crea un pipeline que tiene como pasos, primero preprocessor, donde se escalan los datos, y model. 

Después de esto se crea la variable scores, donde con la función cross_val_score() se llama al pipeline definido, los datos X, los datos y, el tipo de error que se quiere que en este caso es el de neg mean squared error, y el valor de cv de 10.

Se guardan los scores multiplicándolos por menos 1 en la variable mse_scores y se le saca la media y la desviación estándar, que son los valores que se mostrarán al final del código.

# 3

## Correcciones

### 1
1. Quitar las columnas`Gross` y `Runtime` porque no podía trabajar con ellas, y porque se me acababa el tiempo, las quité en lugar de intentar modificarlas para que funcionara.
2. Usar `data['Gross'].str.replace(',', '').astype(float)` y `data['Runtime'] = data['Runtime'].str.replace(' min', '').astype(int)` para hacer que se eliminen las partes del objeto que estaban dando el error cuando las intentaba usar, como "min" y las comas en las cifras de Gross.
3. Cometí el error porque me dio miedo ver el error que arrojó python, y como estaba nerviosa por el examen ni siquiera me tomé algo de tiempo para pensar como solucionarlo, ya que ahora me di cuenta que era muy sencillo y sí sabía hacerlo, pero en el momento ni me pasó por la mente, la verdad.
4. Puedo evitar este error en el futuro intentando no estresarme por un error que no conozco, en lugar de irme al caso extremo y borrar las cosas, puedo intentar solucionarlo de distintas formas.

### 2
1. Me daba este error al hacer la regresión: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names warnings.warn()
2. Lo solucioné usando `data_scaled = data_scaled.to_numpy()`, de esta manera ya no me dio la advertencia.
3. Tampoco me dí cuenta de que me había dado una advertencia, y aunque el solucionarlo el resultado no cambió, prefiero que no me de ninguna advertencia.
4. Puedo evitar este error en el futuro leyendo bien las cosas que hay en mi archivo, ya que aunque esta vez en realidad no cambiaba el resultado, puede haber ocasiones donde sí y es importante darse cuenta.

### 3
1. A la hora de obtener los resultados tenía las variables como x1, x2, ..., x23, lo que no permite saber de qué se está hablando.
2. Usé `data_scaled = pd.DataFrame(data_scaled, columns = x_train.columns)`, para agregar nombres a las columnas, lo que hará que los resultados muestren los nombres de las variables.
3. Me di cuenta de este error, pero en el momento no tuve ninguna idea de como arreglarlo, por lo que mejor continué con el resto del examen, aunque fue muy rápido de solucionar y sí lo pude haber hecho.
4. Puedo evitar este error en el futuro puedo asegurarme de que los DataFrames mantengan los nombres de columnas originales al realizar transformaciones o escalado de datos, lo que facilitará la interpretación de los resultados.

### 4
1. Dije lo que significaba que los p-values fueran menores a 0.05, pero nunca dije qué variables eran significativas.
2. Hice una tabla donde se mostraran solamente las variables que son significativas, donde se viera su coeficiente y el p-value para más facilidad de análisis.
3. Sí pensé en hacerlo, pero mejor me puse a escribir conclusiones, y una vez terminado eso, se me olvidó que iba a arreglar eso y entregué el examen sin corregirlo.
4. Para evitar este error en el futuro puedo revisar bien lo que me pide y asegurarme de haber realizado las cosas que tenía plpaneadas en un principio.

In [265]:
data = pd.read_csv("imdb_top1000_lae.csv")
data = data.drop(columns = ["Unnamed: 0", 'Series_Title', 'Director'])
data = data.dropna()
data.head()

Released_Year Certificate  Runtime  IMDB_Rating  Meta_score        Gross  \
0          1994           A  142 min          9.3        80.0   28,341,469   
1          1972           A  175 min          9.2       100.0  134,966,411   
2          2008          UA  152 min          9.0        84.0  534,858,444   
3          1974           A  202 min          9.0        90.0   57,300,000   
4          1957           U   96 min          9.0        96.0    4,360,000   

   Drama  Crime  Action  Biography  ...  Fantasy  Family  Thriller  Romance  \
0      1      0       0          0  ...        0       0         0        0   
1      1      1       0          0  ...        0       0         0        0   
2      1      1       1          0  ...        0       0         0        0   
3      1      1       0          0  ...        0       0         0        0   
4      1      1       0          0  ...        0       0         0        0   

   Sci-Fi  War  Music  Musical  Sport  History  
0       0    0      0        0      0        0  
1       0    0      0        0      0        0  
2       0    0      0        0      0        0  
3       0    0      0        0      0        0  
4       0    0      0        0      0        0  

[5 rows x 27 columns]

In [266]:
data.describe()

IMDB_Rating  Meta_score       Drama       Crime      Action  \
count   714.000000  714.000000  714.000000  714.000000  714.000000   
mean      7.937115   77.158263    0.700280    0.198880    0.196078   
std       0.293278   12.401144    0.458456    0.399437    0.397307   
min       7.600000   28.000000    0.000000    0.000000    0.000000   
25%       7.700000   70.000000    0.000000    0.000000    0.000000   
50%       7.900000   78.000000    1.000000    0.000000    0.000000   
75%       8.100000   86.000000    1.000000    0.000000    0.000000   
max       9.300000  100.000000    1.000000    1.000000    1.000000   

        Biography     Western      Comedy   Adventure   Animation  ...  \
count  714.000000  714.000000  714.000000  714.000000  714.000000  ...   
mean     0.123249    0.022409    0.225490    0.228291    0.088235  ...   
std      0.328954    0.148113    0.418198    0.420026    0.283836  ...   
min      0.000000    0.000000    0.000000    0.000000    0.000000  ...   
25%      0.000000    0.000000    0.000000    0.000000    0.000000  ...   
50%      0.000000    0.000000    0.000000    0.000000    0.000000  ...   
75%      0.000000    0.000000    0.000000    0.000000    0.000000  ...   
max      1.000000    1.000000    1.000000    1.000000    1.000000  ...   

          Fantasy      Family    Thriller     Romance      Sci-Fi         War  \
count  714.000000  714.000000  714.000000  714.000000  714.000000  714.000000   
mean     0.077031    0.060224    0.138655    0.123249    0.078431    0.040616   
std      0.266827    0.238068    0.345829    0.328954    0.269038    0.197538   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

           Music     Musical       Sport     History  
count  714.00000  714.000000  714.000000  714.000000  
mean     0.04902    0.015406    0.023810    0.053221  
std      0.21606    0.123248    0.152562    0.224632  
min      0.00000    0.000000    0.000000    0.000000  
25%      0.00000    0.000000    0.000000    0.000000  
50%      0.00000    0.000000    0.000000    0.000000  
75%      0.00000    0.000000    0.000000    0.000000  
max      1.00000    1.000000    1.000000    1.000000  

[8 rows x 23 columns]

In [267]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 714 entries, 0 to 997
Data columns (total 27 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Released_Year  714 non-null    object 
 1   Certificate    714 non-null    object 
 2   Runtime        714 non-null    object 
 3   IMDB_Rating    714 non-null    float64
 4   Meta_score     714 non-null    float64
 5   Gross          714 non-null    object 
 6   Drama          714 non-null    int64  
 7   Crime          714 non-null    int64  
 8   Action         714 non-null    int64  
 9   Biography      714 non-null    int64  
 10  Western        714 non-null    int64  
 11  Comedy         714 non-null    int64  
 12  Adventure      714 non-null    int64  
 13  Animation      714 non-null    int64  
 14  Horror         714 non-null    int64  
 15  Mystery        714 non-null    int64  
 16  Film-Noir      714 non-null    int64  
 17  Fantasy        714 non-null    int64  
 18  Family         

In [268]:
data["Released_Year"] = pd.to_numeric(data["Released_Year"], errors='coerce')

In [269]:
data['Runtime'] = data['Runtime'].astype(str)
data['Runtime'] = data['Runtime'].str.replace(' min', '').astype(int)

data['Gross'] = data['Gross'].astype(str)
data['Gross'] = data['Gross'].str.replace(',', '').astype(float)

In [270]:
data

Released_Year Certificate  Runtime  IMDB_Rating  Meta_score        Gross  \
0           1994.0           A      142          9.3        80.0   28341469.0   
1           1972.0           A      175          9.2       100.0  134966411.0   
2           2008.0          UA      152          9.0        84.0  534858444.0   
3           1974.0           A      202          9.0        90.0   57300000.0   
4           1957.0           U       96          9.0        96.0    4360000.0   
..             ...         ...      ...          ...         ...          ...   
990         1971.0          PG      157          7.6        77.0     696690.0   
991         1970.0          GP      144          7.6        50.0    1378435.0   
992         1967.0           U       78          7.6        65.0  141843612.0   
994         1964.0           U       87          7.6        96.0   13780024.0   
997         1953.0      Passed      118          7.6        85.0   30500000.0   

     Drama  Crime  Action  Biography  ...  Fantasy  Family  Thriller  Romance  \
0        1      0       0          0  ...        0       0         0        0   
1        1      1       0          0  ...        0       0         0        0   
2        1      1       1          0  ...        0       0         0        0   
3        1      1       0          0  ...        0       0         0        0   
4        1      1       0          0  ...        0       0         0        0   
..     ...    ...     ...        ...  ...      ...     ...       ...      ...   
990      1      0       0          0  ...        0       0         0        0   
991      0      0       0          0  ...        0       0         0        0   
992      0      0       0          0  ...        0       1         0        0   
994      0      0       0          0  ...        0       0         0        0   
997      1      0       0          0  ...        0       0         0        1   

     Sci-Fi  War  Music  Musical  Sport  History  
0         0    0      0        0      0        0  
1         0    0      0        0      0        0  
2         0    0      0        0      0        0  
3         0    0      0        0      0        0  
4         0    0      0        0      0        0  
..      ...  ...    ...      ...    ...      ...  
990       0    1      0        0      0        0  
991       0    1      0        0      0        0  
992       0    0      0        0      0        0  
994       0    0      1        1      0        0  
997       0    1      0        0      0        0  

[714 rows x 27 columns]

In [271]:
data_dummies = pd.get_dummies(data, columns = ['Certificate'] )
data_dummies = data_dummies.dropna()

In [272]:
x = data_dummies.drop(columns=['IMDB_Rating'])
y = data_dummies['IMDB_Rating']

In [273]:
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.7)

In [274]:
scaler = StandardScaler()
data_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.transform(x_test)
data_scaled = pd.DataFrame(data_scaled, columns = x_train.columns)
data_scaled.head()


Released_Year   Runtime  Meta_score     Gross     Drama     Crime  \
0       1.089938  0.790473    0.540017 -0.658586  0.646230 -0.497494   
1       0.722097  1.056837   -0.088848 -0.586871  0.646230  2.010076   
2       0.301707  0.029435    1.483313 -0.587220 -1.547436 -0.497494   
3       1.037389  0.676318    0.304193 -0.035211  0.646230 -0.497494   
4      -2.168082  3.149693    1.640529 -0.673994  0.646230 -0.497494   

     Action  Biography   Western    Comedy  ...  Certificate_G  \
0 -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
1 -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
2 -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
3 -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
4  2.048882  -0.383571 -0.176045 -0.553428  ...      -0.100605   

   Certificate_GP  Certificate_PG  Certificate_PG-13  Certificate_Passed  \
0       -0.044811       -0.176045          -0.229658            -0.11928   
1       -0.044811       -0.176045          -0.229658            -0.11928   
2       -0.044811       -0.176045          -0.229658            -0.11928   
3       -0.044811       -0.176045          -0.229658            -0.11928   
4       -0.044811       -0.176045          -0.229658            -0.11928   

   Certificate_R  Certificate_TV-PG  Certificate_U  Certificate_U/A  \
0      -0.459527          -0.044811      -0.612091        -0.044811   
1       2.176151          -0.044811      -0.612091        -0.044811   
2      -0.459527          -0.044811       1.633743        -0.044811   
3      -0.459527          -0.044811      -0.612091        -0.044811   
4      -0.459527          -0.044811       1.633743        -0.044811   

   Certificate_UA  
0       -0.491216  
1       -0.491216  
2       -0.491216  
3        2.035763  
4       -0.491216  

[5 rows x 37 columns]

In [275]:
data_scaled = data_scaled.to_numpy()

In [276]:
model = LinearRegression()
model.fit(data_scaled, y_train) 
y_pred_train = model.predict(data_scaled)
r2_train = r2_score(y_train, y_pred_train)
y_pred_test = model.predict(x_test_scaled)
r2_test = r2_score(y_test, y_pred_test)

In [277]:
r2_test, r2_train

(0.18254207112269893, 0.20174874733250503)

In [278]:
data_scaled = pd.DataFrame(data_scaled, columns = x_train.columns)
data_scaled

Released_Year   Runtime  Meta_score     Gross     Drama     Crime  \
0         1.089938  0.790473    0.540017 -0.658586  0.646230 -0.497494   
1         0.722097  1.056837   -0.088848 -0.586871  0.646230  2.010076   
2         0.301707  0.029435    1.483313 -0.587220 -1.547436 -0.497494   
3         1.037389  0.676318    0.304193 -0.035211  0.646230 -0.497494   
4        -2.168082  3.149693    1.640529 -0.673994  0.646230 -0.497494   
..             ...       ...         ...       ...       ...       ...   
494      -0.223780  0.219695   -1.032144  0.054382  0.646230 -0.497494   
495       0.038964  1.323200   -0.088848 -0.640553  0.646230 -0.497494   
496      -1.537497  1.399304    1.011665 -0.622294 -1.547436 -0.497494   
497       1.037389 -0.617448   -0.560496 -0.664506  0.646230 -0.497494   
498       0.617000  1.247097   -0.088848  0.477723  0.646230  2.010076   

       Action  Biography   Western    Comedy  ...  Certificate_G  \
0   -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
1   -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
2   -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
3   -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
4    2.048882  -0.383571 -0.176045 -0.553428  ...      -0.100605   
..        ...        ...       ...       ...  ...            ...   
494 -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
495 -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
496 -0.488071  -0.383571  5.680376 -0.553428  ...      -0.100605   
497 -0.488071  -0.383571 -0.176045 -0.553428  ...      -0.100605   
498 -0.488071   2.607082 -0.176045 -0.553428  ...      -0.100605   

     Certificate_GP  Certificate_PG  Certificate_PG-13  Certificate_Passed  \
0         -0.044811       -0.176045          -0.229658            -0.11928   
1         -0.044811       -0.176045          -0.229658            -0.11928   
2         -0.044811       -0.176045          -0.229658            -0.11928   
3         -0.044811       -0.176045          -0.229658            -0.11928   
4         -0.044811       -0.176045          -0.229658            -0.11928   
..              ...             ...                ...                 ...   
494       -0.044811       -0.176045           4.354308            -0.11928   
495       -0.044811       -0.176045          -0.229658            -0.11928   
496       -0.044811       -0.176045          -0.229658            -0.11928   
497       -0.044811        5.680376          -0.229658            -0.11928   
498       -0.044811       -0.176045          -0.229658            -0.11928   

     Certificate_R  Certificate_TV-PG  Certificate_U  Certificate_U/A  \
0        -0.459527          -0.044811      -0.612091        -0.044811   
1         2.176151          -0.044811      -0.612091        -0.044811   
2        -0.459527          -0.044811       1.633743        -0.044811   
3        -0.459527          -0.044811      -0.612091        -0.044811   
4        -0.459527          -0.044811       1.633743        -0.044811   
..             ...                ...            ...              ...   
494      -0.459527          -0.044811      -0.612091        -0.044811   
495       2.176151          -0.044811      -0.612091        -0.044811   
496      -0.459527          -0.044811      -0.612091        -0.044811   
497      -0.459527          -0.044811      -0.612091        -0.044811   
498      -0.459527          -0.044811      -0.612091        -0.044811   

     Certificate_UA  
0         -0.491216  
1         -0.491216  
2         -0.491216  
3          2.035763  
4         -0.491216  
..              ...  
494       -0.491216  
495       -0.491216  
496       -0.491216  
497       -0.491216  
498       -0.491216  

[499 rows x 37 columns]

In [279]:
ols = sm.OLS(y_train.reset_index(drop=True), sm.add_constant(data_scaled.reset_index(drop=True)))
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            IMDB_Rating   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.140
Method:                 Least Squares   F-statistic:                     3.243
Date:                Sun, 28 Sep 2025   Prob (F-statistic):           4.28e-09
Time:                        21:49:21   Log-Likelihood:                -39.985
No. Observations:                 499   AIC:                             154.0
Df Residuals:                     462   BIC:                             309.8
Df Model:                          36                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                    7.9405      0.012    651.035      0.000       7.917       7.964
Released_Year           -0.0399      0.017     -2.293      0.022      -0.074      -0.006
Runtime                  0.0623      0.015      4.037      0.000       0.032       0.093
Meta_score               0.0673      0.013      5.018      0.000       0.041       0.094
Gross                    0.0366      0.016      2.359      0.019       0.006       0.067
Drama                   -0.0066      0.019     -0.341      0.733      -0.044       0.031
Crime                   -0.0005      0.015     -0.031      0.975      -0.029       0.029
Action                  -0.0083      0.016     -0.522      0.602      -0.040       0.023
Biography               -0.0197      0.014     -1.375      0.170      -0.048       0.008
Western                 -0.0074      0.014     -0.542      0.588      -0.034       0.019
Comedy                  -0.0201      0.016     -1.249      0.212      -0.052       0.012
Adventure               -0.0052      0.018     -0.291      0.771      -0.041       0.030
Animation                0.0067      0.017      0.399      0.690      -0.026       0.040
Horror                  -0.0076      0.014     -0.552      0.581      -0.035       0.019
Mystery                  0.0089      0.014      0.646      0.518      -0.018       0.036
Film-Noir                0.0048      0.013      0.359      0.719      -0.021       0.031
Fantasy                 -0.0023      0.014     -0.171      0.864      -0.029       0.024
Family                  -0.0179      0.014     -1.245      0.214      -0.046       0.010
Thriller                -0.0165      0.014     -1.161      0.246      -0.045       0.011
Romance                 -0.0111      0.014     -0.783      0.434      -0.039       0.017
Sci-Fi                   0.0097      0.014      0.678      0.498      -0.018       0.038
War                     -0.0029      0.013     -0.215      0.830      -0.029       0.023
Music                    0.0297      0.016      1.832      0.068      -0.002       0.062
Musical                 -0.0362      0.016     -2.286      0.023      -0.067      -0.005
Sport                    0.0033      0.013      0.257      0.797      -0.022       0.028
History                 -0.0214      0.014     -1.560      0.119      -0.048       0.006
Certificate_A            0.0200      0.010      1.927      0.055      -0.000       0.040
Certificate_Approved    -0.0030      0.013     -0.226      0.821      -0.029       0.023
Certificate_G            0.0179      0.013      1.365      0.173      -0.008       0.044
Certificate_GP          -0.0110      0.013     -0.857      0.392      -0.036       0.014
Certificate_PG           0.0005      0.012      0.043      0.966      -0.024       0.025
Certificate_PG-13       -0.0196      0.013     -1.547    

In [280]:
tabla = results.summary().tables[1]
df_tabla = pd.DataFrame(tabla.data[1:], columns=tabla.data[0])
df_tabla['P>|t|'] = df_tabla['P>|t|'].astype(float)
tabla_filtrada = df_tabla[df_tabla['P>|t|'] <= 0.05]
tabla_filtrada

coef    std err          t  P>|t|     [0.025  \
0           const      7.9405      0.012    651.035  0.000      7.917   
1   Released_Year     -0.0399      0.017     -2.293  0.022     -0.074   
2         Runtime      0.0623      0.015      4.037  0.000      0.032   
3      Meta_score      0.0673      0.013      5.018  0.000      0.041   
4           Gross      0.0366      0.016      2.359  0.019      0.006   
23        Musical     -0.0362      0.016     -2.286  0.023     -0.067   

       0.975]  
0       7.964  
1      -0.006  
2       0.093  
3       0.094  
4       0.067  
23     -0.005

P values mayores a 0.05: Todos aquellos parámetros cuyos p-values sean vayores a 0.05, significan que su relación con la variable que se esta pronóticando no existe, aunque pueda parecer que sí, por lo que no se tomarán en cuenta estos factores.



El $R^2$ de test da 0.18254207112269893, lo que indica que el modelo no es bueno para predecir la variable objetivo en datos futuros, ya que solo se está explicando el 6% de la varianza del IMBD rating. El $R^2$ de train da 0.20174874733250503, lo que indica que el modelo no tiene un ajuste moderado a los datos de entrenamiento tampoco, pero no es suficiente para garantizar una buena capacidad de generalización.

Las variables que decidí quitar del dataframe original son las siguientes: 
- 'Series_Title' 
- 'Director' 


Decidí quitar los nombres y los directores ya que eran datos que no tenían tanta relevancia para e resultado final, y que en muchos caso no se repetían, por lo que al hacerlas dummies quedaban demasiadas variables.

En conclusión, no resultó un muy buen modelo, como se puede observar con los $R^2$, pero de la información que se obtuvo se puede decir que la mayoría de las variables incluidas tienen poca capacidad explicativa sobre el IMDB rating, el modelo no logra capturar la relación entre las variables predictoras y la variable objetivo, por lo que no es útil para hacer predicciones confiables. Sin embargo, el análisis de los coeficientes y los p-values permite identificar qué variables tienen mayor o menor relevancia estadística, lo que puede orientar futuras mejoras en la selección de variables o en el tipo de modelo utilizado.